### PR TITLE
Support for TawakiV2 (STM32H7)

### PR DESCRIPTION
**Pull request by Fabien-B**

### CODE REVIEW

File: conf/modules/board_tawakiV2.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+
+<module name="board_tawakiV2" dir="boards">
```
Comment (gautierhattenberger) [inline conf/modules/board_tawakiV2.xml:3]: as mentioned, maybe change the name of this module

File: sw/airborne/arch/chibios/mcu_periph/hal_stm32_dma.h [hunk 2]
```diff
@@ -444,7 +448,9 @@ struct DMADriver {
   /**
    * @brief controller associated with stream
    */
+//#if STM32_DMA_SUPPORTS_DMAMUX
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/hal_stm32_dma.h:451]: remove if not needed ?

File: sw/airborne/arch/chibios/mcu_periph/i2c_arch.c [hunk 3]
```diff
@@ -345,8 +345,14 @@ void i2c3_hw_init(void)
 
 #if USE_I2C4
 PRINT_CONFIG_VAR(I2C4_CLOCK_SPEED)
+
+#if defined(STM32H7XX)
+// Local variables (in DMA safe memory)
+static IN_BDMA_SECTION(struct i2c_init i2c4_init_s) = {
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/i2c_arch.c:351]: is this true for all H7 boards ?

**Comment by Fabien-B** [comment]
I will try looking into your suggestions, but it may not be easy.
First annoying thing:

**common_board.h**:
```
#define LED_1_GPIO_ON     gpio_clear
#define LED_1_GPIO_OFF    gpio_set
```
**tawaki_v2.0.h**:
```
#define LED_1_GPIO_ON     gpio_set
#define LED_1_GPIO_OFF    gpio_clear
```

File: sw/airborne/arch/chibios/mcu_periph/spi_arch.c [hunk 4]
```diff
@@ -487,6 +487,31 @@ void spi4_arch_init(void)
 }
 #endif
 
+#if USE_SPI6
+
+#if defined(STM32H7XX)
+// Local variables (in DMA safe memory)
+static IN_BDMA_SECTION(struct spi_init spi6_init_s) = {
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/spi_arch.c:494]: same question. Is it for all H7 boards ? only for Tawaki ? what about the other SPI or I2C peripherals ?

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

File: sw/airborne/arch/chibios/mcu_periph/i2c_arch.c (hunk 3 above)
Comment (Fabien-B) [reply sw/airborne/arch/chibios/mcu_periph/i2c_arch.c:351]: yes

File: sw/airborne/arch/chibios/mcu_periph/spi_arch.c (hunk 4 above)
Comment (Fabien-B) [reply sw/airborne/arch/chibios/mcu_periph/spi_arch.c:494]: yes, SPI6 is on BDMA.

**Comment by alex31** [comment]
> What about the defines in `mcuconf.h` is this not enough for you to set the nocache regions?:
> 
> ```
>  * Memory attributes settings.
>  */
> #define STM32_NOCACHE_ENABLE                TRUE
> #define STM32_NOCACHE_MPU_REGION            MPU_REGION_6
> #define STM32_NOCACHE_RBAR                  0x24000000U
> #define STM32_NOCACHE_RASR                  MPU_RASR_SIZE_64K
> ```
> 
> Because I prefer to keep it then the same for every board, either we do it ourselves or use the ChibiOS features to set the MPU.
> 
> Next to that your [tawaki_v2.0.h](https://github.com/paparazzi/paparazzi/pull/3199/files#diff-75a9a301594643dbefcdd4ea7ecc73e73135ef3ea9e6d6818a2a17058339a254) is almost the same as the `common_board.h`. Isn't it better to change the names in `board.cfg` such that `common_board.h` can be used and the features that are missing can be added there?
> 
> These are just some suggestions, but I really like the fixes that you have done :)

The H7 provides 3 different domains, and we need "nocache" regions in 2 regions, to support all of our dma bound peripherals. Chibios mechanism permit to define only one region via mcuconf.h.  
We define 2 regions in our custom code, so STM32_NOCACHE_ENABLE should be set to FALSE in mcuconf.h

File: sw/airborne/arch/chibios/mcu_periph/i2c_arch.c (hunk 3 above)
Comment (alex31) [reply sw/airborne/arch/chibios/mcu_periph/i2c_arch.c:351]: yes I2C4 and SPI6 are peripherals linked to BDMA on all STM32H7 whatever the card that use them.
the rational behind the use of I2C4 and SPI6 on the tawakiV2 is to migrate maximum number of peripherals from DMA to BDMA to free DMA channels that we want to later use with BIDIR DSHOT which need N+1 DMA channels for each timer using Bidir DSHOT, so up to 10 DMA channels for octocopter.

File: sw/airborne/arch/chibios/mcu_periph/spi_arch.c (hunk 4 above)
Comment (alex31) [reply sw/airborne/arch/chibios/mcu_periph/spi_arch.c:494]: cf I2C4 comment

File: conf/airframes/ENAC/quadrotor/TawakiV2.xml [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    <target name="ap" board="tawaki_2.0">
+      <define name="RADIO_KILL_SWITCH" value="RADIO_GAIN1"/>
+    </target>
+
+    <target name="nps" board="pc">
+      <module name="fdm" type="jsbsim"/>
+      <module name="radio_control" type="ppm"/>
+    </target>
+
+    <module name="board" type="tawakiV2">
```
[the remark this answers is not in the document]
Comment (Fabien-B) [reply conf/airframes/ENAC/quadrotor/TawakiV2.xml:23]: There is `boards/tawaki2.0.xml` which is different from `board_tawakiV2.xml`.
I tried reproducing the same files as for Tawaki.
What should I do ?

**Review by gautierhattenberger — APPROVED** [review]
Can be merged once tested in flight

**Comment by Fabien-B** [comment]
Tested multiple times, it works.

**Pull request merged**